python/thunder/imgprocessing/regmethods/utils.py

```diff
@@ -97,4 +97,165 @@ def computeDisplacement(arry1, arry2):
     pairs = zip(maxinds, arry1.shape)
     adjusted = [d - n if d > n // 2 else d for (d, n) in pairs]
 
-    return adjusted
\ No newline at end of file
+    return adjusted
+
+
+def zeroBorder(vol, border=1, cval=0.0):
+    """Zero-out boundary voxels of a volume.
+
+    Parameters
+    ----------
+        vol: 3d volume
+        border: scalar or tuple containing borders for each dimension
+        cval: constant value to replace boundary voxels with, default is 0.0
+    """
+    import numpy as np # sorry
+    vol = vol.copy() # create new copy so we don't overwrite vol
+    dims = np.array(vol.shape)
+    if np.size(border) == 1:
+        border = border * np.ones(vol.ndim, dtype=int)
+    border = np.array(border, dtype=int)
+    # Don't apply border to singleton dimensions.
+    border[dims == 1] = 0
+    assert len(border) == vol.ndim
+    if np.any(dims - border <= 0):
+        raise ValueError('Border %s exceeds volume shape %s.' %
+                (str(border), str(dims)) )
+    for dim, bval in enumerate(border):
+        if bval > 0:
+            slices = [slice(bval) if d == dim else slice(None) for d in xrange(vol.ndim)]
+            vol[slices] = cval
+            slices[dim] = slice(-bval, None)
+            vol[slices] = cval
+    return vol
+
+
+def solveLinearized(vec, jacobian, reference, robust=False):
+    """Solve linearized registration problem for change in transformation parameters and weighting on reference basis.
+
+    Parameters
+    ----------
+    vec : array, shape (nvoxels,)
+        vectorized volume
+    jacobian: array, shape (nvoxels, nparams)
+        jacobian for each transformation parameter
+    reference: array, shape (nvoxels, nbasis)
+        array of vectorized reference volumes
+    robust: bool, optional, default = False
+        solve a least absolute deviation problem instead of least squares
+
+    Returns
+    -------
+    deltaTransforms : array, shape (nparams,)
+        optimal change in transformation parameters
+    coeff : array, shape (nbasis,)
+        optimal weighting of reference volumes
+    """
+    from numpy import column_stack
+    A = column_stack((jacobian, reference))
+    if robust:
+        from statsmodels.regression.quantile_regression import QuantReg
+        quantile = 0.5
+        model = QuantReg(vec, A).fit(q=quantile)
+        params = model.params
+    else:
+        from numpy.linalg import lstsq
+        model = lstsq(A, vec)
+        params = model[0]
+    from numpy import split
+    deltaTransform, coeff = split(params, [jacobian.shape[1]])
+    return deltaTransform, coeff
+
+
+def v2v(v, dims=None):
```

Comment:
Maybe it would be cleaner / clearer to break this into two functions, one that goes one way and the other going the other way?